### PR TITLE
fix(client): Errors are obfuscated by interactive transactions gh-12862

### LINKED ISSUE #12862: Some errors are obfuscated by interactive transactions when using `binary` engine
### Bug description

Runtime errors such as `CHECK` constraint failures are obfuscated by interactive transactions when `engineType  = "binary"`, making it impossible to appropriately handle rejected transactions.

Only tested this with `postgres` so far. 

Note that the `library` engine does not seem to have the same issue. CI output comparing both engine types: https://github.com/driimus/prisma-interactive-tx-errors/actions/runs/2184082955

### How to reproduce

1. Clone https://github.com/driimus/prisma-interactive-tx-errors
2. Run the tests locally using a `binary` engine
3. Notice that the test using an interactive transaction fails due to a nondescript error.

### Expected behavior

The original error / error message should be surfaced, similar to the `library` engine's behavior.

Expected:

```
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: \"ERROR\", parsed_severity: Some(Error), code: SqlState(\"23514\"), message: \"new row for relation \\\"Post\\\" violates check constraint \\\"Post_viewCount_check\\\"\", detail: Some(\"Failing row contains (1, 2022-04-18 12:42:07.239, 2022-04-18 12:42:07.239, Minima nulla ea., null, f, -1, 1).\"), hint: None, position: None, where_: None, schema: Some(\"public\"), table: Some(\"Post\"), column: None, datatype: None, constraint: Some(\"Post_viewCount_check\"), file: Some(\"execMain.c\"), line: Some(2003), routine: Some(\"ExecConstraints\") }) }) })"
```

Received:

```
 ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: \"ERROR\", parsed_severity: Some(Error), code: SqlState(\"25P02\"), message: \"current transaction is aborted, commands ignored until end of transaction block\", detail: None, hint: None, position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some(\"postgres.c\"), line: Some(1718), routine: Some(\"exec_bind_message\") }) }) })"
```

### Prisma information

[Minimal reproduction repo](https://github.com/driimus/prisma-interactive-tx-errors)

```schema.prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["interactiveTransactions"]
  engineType      = "binary"
}

datasource db {
  provider = "postgres"
  url      = "postgresql://postgres:example@localhost"
}

model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String?
  posts Post[]
}

model Post {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  title     String
  content   String?
  published Boolean  @default(false)
  viewCount Int      @default(0)
  author    User?    @relation(fields: [authorId], references: [id])
  authorId  Int?
}
```

### Environment & setup

- OS: `Ubuntu 20.04.4 LTS`
- Database: `PostgreSQL`
- Node.js version: `16.14.2`

### Prisma Version

```
prisma                  : 3.12.0
@prisma/client          : 3.12.0
Current platform        : debian-openssl-1.1.x
Query Engine (Node-API) : libquery-engine 22b822189f46ef0dc5c5b503368d1bee01213980 (at node_modules/.pnpm/@prisma+engines@3.12.0-37.22b822189f46ef0dc5c5b503368d1bee01213980/node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Migration Engine        : migration-engine-cli 22b822189f46ef0dc5c5b503368d1bee01213980 (at node_modules/.pnpm/@prisma+engines@3.12.0-37.22b822189f46ef0dc5c5b503368d1bee01213980/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine    : introspection-core 22b822189f46ef0dc5c5b503368d1bee01213980 (at node_modules/.pnpm/@prisma+engines@3.12.0-37.22b822189f46ef0dc5c5b503368d1bee01213980/node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary           : prisma-fmt 22b822189f46ef0dc5c5b503368d1bee01213980 (at node_modules/.pnpm/@prisma+engines@3.12.0-37.22b822189f46ef0dc5c5b503368d1bee01213980/node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash    : 22b822189f46ef0dc5c5b503368d1bee01213980
Studio                  : 0.459.0
```

### PR DESCRIPTION
**Pull request by danstarns**
Closes: https://github.com/prisma/prisma/issues/12862

### CODE REVIEW

File: packages/client/tests/functional/_utils/setupTestSuiteEnv.ts [hunk 1]
```diff
@@ -106,12 +107,30 @@ export async function setupTestSuiteDatabase(
   suiteMeta: TestSuiteMeta,
   suiteConfig: NamedTestSuiteConfig,
   errors: Error[] = [],
+  alterStatement?: string,
```
Comment (millsp) [inline packages/client/tests/functional/_utils/setupTestSuiteEnv.ts:110]: Could we have an alter statement per provider? Could be useful for testing similar things on all SQL providers

File: packages/client/tests/functional/_utils/setupTestSuiteEnv.ts [hunk 2]
```diff
@@ -106,12 +107,30 @@ export async function setupTestSuiteDatabase(
...
   try {
     const consoleInfoMock = jest.spyOn(console, 'info').mockImplementation()
     await DbPush.new().parse(['--schema', schemaPath, '--force-reset', '--skip-generate'])
+
+    if (alterStatement) {
+      const prismaDir = path.dirname(schemaPath)
+      const timestamp = new Date().getTime()
+
+      await fs.promises.mkdir(`${prismaDir}/migrations/${timestamp}`, { recursive: true })
+      await fs.promises.writeFile(`${prismaDir}/migrations/migration_lock.toml`, 'provider = "postgresql"')
```
Comment (millsp) [inline packages/client/tests/functional/_utils/setupTestSuiteEnv.ts:123]: We should use the provider dynamically here, I think? Is this file really needed?

**Comment by danstarns** [comment]
The best way to integrate and test an engines fix for this is to use the `enginesOverride` in your equivalent of: 

```
/Users/danielstarns/code/clones/prisma/packages/fetch-engine/package.json
``` 

```diff
  "repository": {
    "type": "git",
    "url": "https://github.com/prisma/prisma.git",
    "directory": "packages/fetch-engine"
  },
  "bugs": "https://github.com/prisma/prisma/issues",
+  "enginesOverride": {
+    "folder": "/Users/danielstarns/code/clones/prisma-engines/target/release"
+  },
  "devDependencies": {
```

You can then build your rust branch that contains the fix, and do an `pnpm i` in the **root** of `prisma/prisma`. 

Finally, you can now run the test suite that is in this branch. 

```
cd ./packages/client
PRISMA_CLIENT_ENGINE_TYPE="binary" pnpm run test:functional:code 12862
```

You should expect one test to fail until the fix is applied to engines. 


Current: 

```
FAIL  tests/functional/issues/12862/tests.ts
  issues.12862 (provider=postgresql)
    ✓ should propagate the correct error when a method fails (339 ms)
    ✓ should propagate the correct error when a method fails inside an transaction (19 ms)
    ✕ should propagate the correct error when a method fails inside an interactive transaction (173 ms)

  ● issues.12862 (provider=postgresql) › should propagate the correct error when a method fails inside an interactive transaction

    expect(received).rejects.toThrowError(expected)

    Expected substring: "violates check constraint \\\"post_viewcount_check\\\""
    Received message:   "
    Invalid `client.post.create()` invocation in
    /Users/danielstarns/code/clones/prisma/packages/client/tests/functional/issues/12862/tests.ts:63:42·
      60   },
      61 })
      62·
    → 63 const post = await client.post.create(
    Error occurred during query execution:
    ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: \"ERROR\", parsed_severity: Some(Error), code: SqlState(E25P02), message: \"current transaction is aborted, commands ignored until end of transaction block\", detail: None, hint: None, position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some(\"postgres.c\"), line: Some(1591), routine: Some(\"exec_bind_message\") }) }) })"

          30853 |       throw new PrismaClientRustPanicError(message, this.client._clientVersion);
          30854 |     } else if (error2 instanceof PrismaClientUnknownRequestError) {
        > 30855 |       throw new PrismaClientUnknownRequestError(message, this.client._clientVersion);
                |             ^
          30856 |     } else if (error2 instanceof PrismaClientInitializationError) {
          30857 |       throw new PrismaClientInitializationError(message, this.client._clientVersion);
          30858 |     } else if (error2 instanceof PrismaClientRustPanicError) {

          at RequestHandler.handleRequestError (runtime/index.js:30855:13)
          at RequestHandler.handleRequestError [as request] (runtime/index.js:30834:12)
          at PrismaClient._request (runtime/index.js:31812:16)
          at tests/functional/issues/12862/tests.ts:63:24
          at Proxy._transactionWithCallback (runtime/index.js:31744:18)
          at Object.<anonymous> (tests/functional/issues/12862/tests.ts:54:7)

      71 |           return post
      72 |         }),
    > 73 |       ).rejects.toThrowError('violates check constraint \\"post_viewcount_check\\"')
         |                 ^
      74 |     })
      75 |   },
      76 |   {

      at Object.toThrowError (../../../../node_modules/.pnpm/expect@28.1.3/node_modules/expect/build/index.js:241:22)
      at Object.toThrowError (issues/12862/tests.ts:73:17)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 2 passed, 3 total
Snapshots:   0 total
Time:        4.271 s, estimated 8 s
Ran all test suites matching /12862/i.
 ELIFECYCLE  Command failed with exit code 1.
```


Expected


```
PASS  tests/functional/issues/12862/tests.ts
  issues.12862 (provider=postgresql)
    ✓ should propagate the correct error when a method fails (339 ms)
    ✓ should propagate the correct error when a method fails inside an transaction (19 ms)
    ✓ should propagate the correct error when a method fails inside an interactive transaction (173 ms)
```

**Comment by garrensmith** [comment]
This issue is in the BinaryEngine in the client. On this line here:
https://github.com/prisma/prisma/blob/fix/12862/packages/engine-core/src/binary/BinaryEngine.ts#L970

We will retry a failed operation. This leads to the interactive transaction returning a different unexpected error than the actual error that caused the problem. 

I'm not sure why we retry operations in the binary engine. I think this is a bad idea and we should not retry at all. 
If there is a good reason for that, then we should not retry interactive transaction operations.

File: packages/client/tests/functional/_utils/setupTestSuiteEnv.ts (hunk 2 above)
Comment (danstarns) [reply packages/client/tests/functional/_utils/setupTestSuiteEnv.ts]: Sure!

File: packages/client/tests/functional/_utils/setupTestSuiteEnv.ts (hunk 1 above)
Comment (danstarns) [reply packages/client/tests/functional/_utils/setupTestSuiteEnv.ts]: I have made it a callback where the provider is passed in.

**Comment by danstarns** [comment]
Waiting for merge of https://github.com/prisma/prisma/pull/15684

**Comment by danstarns** [comment]
Merged into https://github.com/prisma/prisma/pull/15684
